### PR TITLE
create map

### PR DESCRIPTION
**Pull request by sebakocz**

### CODE REVIEW

**Comment by sebakocz** [comment]
[create test map](https://trello.com/c/2Z8sqen7/7-create-test-map)

File: src/config.ts [hunk 1]
```diff
@@ -4,16 +4,17 @@ import MainScene from './scenes/mainScene';
...
   scale: {
     // has to be aspect ratio of 2:1
     width: 1000,
     height: 500,
   },
   physics: {
-    default: 'arcade',
-    arcade: {
-      gravity: { y: 200 },
+    default: 'matter',
```
Comment (sebakocz) [inline src/config.ts:14]: is apparently the better cooler mode

File: src/config.ts [hunk 2]
```diff
@@ -4,16 +4,17 @@ import MainScene from './scenes/mainScene';
...
+    matter: {
+      gravity: { y: 1 },
+      // debug: true,
```
Comment (sebakocz) [inline src/config.ts:17]: flip this on and off to see bounding boxes for collisions, nice for debugging

File: test.html
Comment (sebakocz) [inline test.html:1]: wanted to have an entry point for faster feedback loops during testing

this is not included in the build 🤓 👍

File: src/scenes/mainScene.ts
Comment (sebakocz) [inline src/scenes/mainScene.ts:1]: this file is pure magic, pls be gentle

https://labs.phaser.io/ is a great place for resources, I stole the spline code from there for chad-gpt driven development

**Review by sebakocz — COMMENTED** [review]
added own thoughts

File: src/config.ts (hunk 2 above)
Comment (nils-keller-dev) [reply src/config.ts:17]: image
Is this an easy fix?

File: src/scenes/mainScene.ts [hunk 3]
```diff
@@ -1,25 +1,131 @@
...
+    graphics.fillStyle(GROUND_COLOR);
+    this.createGroundPolygons(hills, lineCategory);
+    this.drawAndFillGroundCurve(graphics, curve);
+    this.createBall();
   }
 
-  create() {
-    // Set up physics and enable gravity in the y-axis
-    this.physics.world.gravity.y = 300;
+  createGroundCurve(hills: { x: number; y: number }[]) {
```
Comment (nils-keller-dev) [inline src/scenes/mainScene.ts]: new interface for `{ x: number; y: number }`?
can be located at the start of this file

File: src/scenes/mainScene.ts [hunk 4]
```diff
@@ -1,25 +1,131 @@
...
+    const stiffness = 0.1;
+    const options = {
+      friction: 0,
+      frictionAir: 0,
+      restitution: 0,
+      ignoreGravity: true,
+      inertia: Infinity,
+      isStatic: true,
+      angle: 0,
+      collisionFilter: { category: lineCategory },
```
Comment (nils-keller-dev) [inline src/scenes/mainScene.ts]:
```suggestion
      collisionFilter: { category },
```

File: src/scenes/mainScene.ts [hunk 5]
```diff
@@ -1,25 +1,131 @@
...
-    // Set up physics and enable gravity in the y-axis
-    this.physics.world.gravity.y = 300;
+  createGroundCurve(hills: { x: number; y: number }[]) {
+    const curvePoints = hills.flatMap((hill) => [hill.x, hill.y]);
+    return new Phaser.Curves.Spline(curvePoints);
+  }
+
+  createGroundPolygons(
+    hills: { x: number; y: number }[],
+    lineCategory: number
```
Comment (nils-keller-dev) [inline src/scenes/mainScene.ts]:
```suggestion
    category: number
```

File: src/scenes/mainScene.ts [hunk 6]
```diff
@@ -1,25 +1,131 @@
...
+        endPoint.y
+      );
+
+      const previous = this.matter.add.polygon(
+        startPoint.x,
+        startPoint.y,
+        sides,
+        size,
+        options
+      );
```
Comment (nils-keller-dev) [inline src/scenes/mainScene.ts:75]:
```suggestion
      );
      
```

File: src/scenes/mainScene.ts [hunk 7]
```diff
@@ -1,25 +1,131 @@
...
+        Math.random() * hillHeight * randomFactor -
+        (hillHeight * randomFactor) / 2;
+      hills.push(new Phaser.Math.Vector2(x, y));
+    }
 
-    ball.setCollideWorldBounds(true);
-    ball.setBounce(1, 1);
-    ball.setVelocity(Phaser.Math.Between(-200, 200), 20);
+    return hills;
   }
```
Comment (nils-keller-dev) [inline src/scenes/mainScene.ts]: we should move this function outside this file in the long term, once multiple players online need to see the same map.
Maybe it makes sense to do this now, but you can decide

File: src/scenes/mainScene.ts [hunk 8]
```diff
@@ -1,25 +1,131 @@
...
-    ball.setScale(0.2);
+    for (let i = 0; i < hillCount; i++) {
+      const x = (width / (hillCount - 1)) * i;
+      const y =
+        hillHeight +
+        (Math.sin((Math.PI / (hillCount - 1)) * i) * hillHeight) / 2 +
+        Math.random() * hillHeight * randomFactor -
+        (hillHeight * randomFactor) / 2;
+      hills.push(new Phaser.Math.Vector2(x, y));
+    }
```
Comment (nils-keller-dev) [inline src/scenes/mainScene.ts]:
```suggestion
    return Array.from({ length: hillCount }).map((e, i) => {
      const x = (width / (hillCount - 1)) * i;
      const y =
        hillHeight +
        (Math.sin((Math.PI / (hillCount - 1)) * i) * hillHeight) / 2 +
        Math.random() * hillHeight * randomFactor -
        (hillHeight * randomFactor) / 2;
      return new Phaser.Math.Vector2(x, y);
    });
```

File: src/scenes/mainScene.ts [hunk 9]
```diff
@@ -1,25 +1,131 @@
...
+    this.matter.add.gameObject(ball, {
+      shape: 'circle',
+      radius: 50,
+      restitution: 0.1,
+    });
+  }
 
-    const ball = this.physics.add.image(400, 300, 'ball');
+  generateHills(width: number, height: number) {
+    const hills = [];
```
Comment (nils-keller-dev) [inline src/scenes/mainScene.ts]:
```suggestion
```

File: src/scenes/mainScene.ts [hunk 10]
```diff
@@ -1,25 +1,131 @@
...
+        (Math.sin((Math.PI / (hillCount - 1)) * i) * hillHeight) / 2 +
+        Math.random() * hillHeight * randomFactor -
+        (hillHeight * randomFactor) / 2;
+      hills.push(new Phaser.Math.Vector2(x, y));
+    }
 
-    ball.setCollideWorldBounds(true);
-    ball.setBounce(1, 1);
-    ball.setVelocity(Phaser.Math.Between(-200, 200), 20);
+    return hills;
```
Comment (nils-keller-dev) [inline src/scenes/mainScene.ts]:
```suggestion
```

File: test.html [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+            name="viewport"
+            content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0"
+    />
+    <link rel="stylesheet" href="/src/css/styles.css" />
+    <script type="module" src="/src/index.ts"></script>
+</head>
+<body>
+<div id="game"></div>
+</body>
+</html>
```
Comment (nils-keller-dev) [inline test.html]: dont you have prettier installed properly?

**Comment by nils-keller-dev** [comment]
I love this, it looks amazing!

**Review by nils-keller-dev — CHANGES_REQUESTED** [review]

File: test.html (hunk 11 above)
Comment (sebakocz) [reply test.html]: ye you were right, fixed with next commit, I guess my IDE didn't like that there was no .prettierrc

File: src/scenes/mainScene.ts (hunk 8 above)
Comment (sebakocz) [reply src/scenes/mainScene.ts]: nice ye looks smooth af

File: src/scenes/mainScene.ts (hunk 7 above)
Comment (sebakocz) [reply src/scenes/mainScene.ts]: will move this elsewhere

ye was thinking to use seeds to make sure everyone gets the same map

or additionally provide various different inputs so we can tweak how crazy the amplitudes go

File: src/scenes/mainScene.ts (hunk 6 above)
Comment (sebakocz) [reply src/scenes/mainScene.ts:75]: ok ok

File: src/scenes/mainScene.ts (hunk 4 above)
Comment (sebakocz) [reply src/scenes/mainScene.ts]: ah nice catch!

File: src/scenes/mainScene.ts (hunk 3 above)
Comment (sebakocz) [reply src/scenes/mainScene.ts]: ye makes sense! 👍

File: src/config.ts (hunk 2 above)
Comment (sebakocz) [reply src/config.ts:17]: no but magic happened, 
I also reduced the amount of segments cause so many are not needed

**Pull request merged**